pkgs/development/libraries/openmpi/default.nix

```diff
@@ -1,136 +1,270 @@
+      # are 7x3 such options. We use lib.mapCartesianProduct to iterate them all.
+      fileNamesToIterate = {
+        p = [
+          "mpi"
+          "shmem"
+          "osh"
+        ];
+        s =
+          [
+            "CC"
+            "c++"
+            "cxx"
+            "cc"
+          ]
+          ++ lib.optionals fortranSupport [
+            "f77"
+            "f90"
+            "fort"
+          ];
+      };
+      wrapperDataSubstitutions =
+        {
+          # The attr key is the filename prefix. The list's 1st value is the
+          # compiler=_ line that should be replaced by a compiler=#2 string, where
+          # #2 is the 2nd value in the list.
+          "cc" = [
+            "gcc"
+            "${targetPackages.stdenv.cc}/bin/${targetPackages.stdenv.cc.targetPrefix}cc"
+          ];
+          "c++" = [
+            "g++"
+            "${targetPackages.stdenv.cc}/bin/${targetPackages.stdenv.cc.targetPrefix}c++"
+          ];
+        }
+        // lib.optionalAttrs fortranSupport {
+          "fort" = [
+            "gfortran"
+            "${targetPackages.gfortran}/bin/${targetPackages.gfortran.targetPrefix}gfortran"
+          ];
+        };
+      # The -wrapper-data.txt files that are not symlinks, need to be iterated as
+      # well, here they start withw ${part1}${part2}, and we use
+      # lib.mapCartesianProduct as well.
+      wrapperDataFileNames = {
+        part1 = [
+          "mpi"
+          "shmem"
+        ];
+        part2 = builtins.attrNames wrapperDataSubstitutions;
+      };
+    in
+    ''
+      find $out/lib/ -name "*.la" -exec rm -f \{} \;
 
-    # default compilers should be indentical to the
-    # compilers at build time
+      # The main wrapper that all the rest of the commonly used binaries are
+      # symlinked to
+      moveToOutput "bin/opal_wrapper" "''${!outputDev}"
+      # All of the following files are symlinks to opal_wrapper
+      ${lib.pipe fileNamesToIterate [
+        (lib.mapCartesianProduct (
+          { p, s }:
+          ''
+            echo "handling ${p}${s}"
+            moveToOutput "bin/${p}${s}" "''${!outputDev}"
+            moveToOutput "share/openmpi/${p}${s}-wrapper-data.txt" "''${!outputDev}"
+          ''
+        ))
+        (lib.concatStringsSep "\n")
+      ]}
+      # default compilers should be indentical to the
+      # compilers at build time
+      ${lib.pipe wrapperDataFileNames [
+        (lib.mapCartesianProduct (
+          { part1, part2 }:
+          ''
+            substituteInPlace "''${!outputDev}/share/openmpi/${part1}${part2}-wrapper-data.txt" \
+              --replace-fail \
+                compiler=${lib.elemAt wrapperDataSubstitutions.${part2} 0} \
+                compiler=${lib.elemAt wrapperDataSubstitutions.${part2} 1}
+          ''
+        ))
+        (lib.concatStringsSep "\n")
+      ]}
+      # A symlink to ${lib.getDev pmix}/bin/pmixcc upstreeam puts here as well
```

Comment:
This comment gets interpolated and therefore requires `pmix` which is not available on Darwin.

Reply:
> This comment gets interpolated and therefore requires `pmix` which is not available on Darwin.

pmix is available on Darwin since https://github.com/NixOS/nixpkgs/pull/333016 .